src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/LocalDB.Windows.cs

```diff
@@ -0,0 +1,222 @@
+
+        // Local Db api doc https://msdn.microsoft.com/en-us/library/hh217143.aspx
+        // HRESULT LocalDBStartInstance( [Input ] PCWSTR pInstanceName, [Input ] DWORD dwFlags,[Output] LPWSTR wszSqlConnection,[Input/Output] LPDWORD lpcchSqlConnection);  
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        internal delegate int LocalDBStartInstance(
+                [In] [MarshalAs(UnmanagedType.LPWStr)] string localDBInstanceName,
+                [In]  int flags,
+                [Out] [MarshalAs(UnmanagedType.LPWStr)] StringBuilder sqlConnectionDataSource,
+                [In, Out]ref int bufferLength);
+
+        private LocalDBStartInstance localDBStartInsanceFunc = null;
+
+        private volatile SafeLibraryHandle _sqlUserInstanceLibraryHandle;
+
+        private LocalDB() { }
+
+        internal static string GetLocalDBConnectionString(string localDbInstance) =>
+            Instance.LoadUserInstanceDll() ? Instance.GetConnectionString(localDbInstance) : null;
+
+        private string GetConnectionString(string localDbInstance)
+        {
+            StringBuilder localDBConnectionString = new StringBuilder(MAX_LOCAL_DB_CONNECTION_STRING_SIZE + 1);
+            int sizeOfbuffer = localDBConnectionString.Capacity;
+            localDBStartInsanceFunc(localDbInstance, 0, localDBConnectionString, ref sizeOfbuffer);
+            return localDBConnectionString.ToString();
+        }
+
+        internal enum LocalDBErrorState
+        {
+            NO_INSTALLATION, INVALID_CONFIG, NO_SQLUSERINSTANCEDLL_PATH, INVALID_SQLUSERINSTANCEDLL_PATH, NONE
+        }
+
+        internal static uint MapLocalDBErrorStateToCode(LocalDBErrorState errorState)
+        {
+            switch (errorState)
+            {
+                case LocalDBErrorState.NO_INSTALLATION:
+                    return SNICommon.LocalDBNoInstallation;
+                case LocalDBErrorState.INVALID_CONFIG:
+                    return SNICommon.LocalDBInvalidConfig;
+                case LocalDBErrorState.NO_SQLUSERINSTANCEDLL_PATH:
+                    return SNICommon.LocalDBNoSqlUserInstanceDllPath;
+                case LocalDBErrorState.INVALID_SQLUSERINSTANCEDLL_PATH:
+                    return SNICommon.LocalDBInvalidSqlUserInstanceDllPath;
+                case LocalDBErrorState.NONE:
+                    return 0;
+                default:
+                    return SNICommon.LocalDBInvalidConfig;
+            }
+        }
+
+        /// <summary>
+        /// Loads the User Instance dll.
+        /// </summary>
+        private bool LoadUserInstanceDll()
+        {
+            // Check in a non thread-safe way if the handle is already set for performance.
+            if (_sqlUserInstanceLibraryHandle != null)
+            {
+                return true;
+            }
+
+            lock (this)
+            {
+                if(_sqlUserInstanceLibraryHandle !=null)
+                {
+                    return true;
+                }
```

Comment:
You can use LazyInitializer.EnsureInitalized instead of doing the double-checked locking yourself.